debian/sourcemapper.go

```diff
@@ -0,0 +1,161 @@
+// to populate and update the map. It is Release-centric.
+//
+// It should have the same lifespan as the Updater to save allocations
+// and take advantage of the entity tag that debian sends back.
+type SourcesMap struct {
+	release    Release
+	sourcesURL string
+	mu, etagMu *sync.RWMutex
+	sourceMap  map[string]map[string]struct{}
+	etagMap    map[string]string
+	client     *http.Client
+}
+
+// Get returns all the binaries associated with a source package
+// identified by a string. Empty slice is returned if the source
+// doesn't exist in the map.
+func (m *SourcesMap) Get(source string) []string {
+	if m.sourceMap[source] == nil {
+		return []string{}
+	}
+	bins := []string{}
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for bin := range m.sourceMap[source] {
+		bins = append(bins, bin)
+	}
+	return bins
+}
+
+// Update pulls the Sources.gz files for the different repos and saves
+// the resulting source to binary relationships.
+func (m *SourcesMap) Update(ctx context.Context) error {
+	if m.release == Wheezy {
+		// There are no Wheezy records we assume the source->binary relationship of Jessie.
+		m.release = Jessie
+	}
+	g, ctx := errgroup.WithContext(ctx)
+	for _, r := range repos {
+		url := fmt.Sprintf(m.sourcesURL, m.release, r)
+		g.Go(func() error {
+			err := m.fetchSources(ctx, url)
+			if err != nil {
+				return fmt.Errorf("unable to fetch sources: %w", err)
+			}
+			return nil
+		})
+	}
+	return g.Wait()
+}
+
+func (m *SourcesMap) fetchSources(ctx context.Context, url string) error {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "debian/sourcemapper.fetchSources"),
+		label.String("url", url))
+	zlog.Debug(ctx).Msg("attempting fetch of Sources file")
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+	m.etagMu.RLock()
+	req.Header.Set("If-None-Match", m.etagMap[url])
+	m.etagMu.RUnlock()
+
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusOK:
+	case http.StatusNotModified:
+		zlog.Debug(ctx).Msg("already processed the latest version of the file")
+		return nil
+	default:
+		return fmt.Errorf("received status code %d querying mapping url %s", resp.StatusCode, url)
+	}
+	m.etagMu.Lock()
+	m.etagMap[url] = resp.Header.Get("etag")
+	m.etagMu.Unlock()
+
+	var reader io.ReadCloser
+	switch resp.Header.Get("Content-Type") {
+	case "application/gzip", "application/x-gzip":
+		reader, err = gzip.NewReader(resp.Body)
+		if err != nil {
+			return err
+		}
+		defer reader.Close()
+	default:
+		return fmt.Errorf("received bad content-type %s querying mapping url %s", resp.Header.Get("Content-Type"), url)
+	}
+
+	tp := textproto.NewReader(bufio.NewReader(reader))
+	hdr, err := tp.ReadMIMEHeader()
+	for ; err == nil && len(hdr) > 0; hdr, err = tp.ReadMIMEHeader() {
+		source := hdr.Get("Package")
+		if source == "linux" {
```

Comment:
Does this ever have the version attached? I know it sometimes happens in other apt databases.

Reply:
There are some binaries with the version attached but it seems like they tend to be for the packages that also have the package attatched i.e.: 

```
Package: tcl8.6
Binary: tcl8.6, libtcl8.6, tcl8.6-dev, tcl8.6-doc
Version: 8.6.11+dfsg-1
Maintainer: Debian Tcl/Tk Packagers <pkg-tcltk-devel@lists.alioth.debian.org>
```

Reply:
Yeah, poking around, it seems to only be in the `Packages` file, for when multiple builds were needed. E.g.

```Packages
Package: acmetool
Source: acmetool (0.2.1-4)
Version: 0.2.1-4+b5
```
and
```Sources
Package: acmetool
Binary: acmetool
Version: 0.2.1-4
```